plugin/PlayAssetDelivery/src/main/java/com/google/play/core/godot/assetpacks/utils/AssetPackStateFromDictionary.java

```diff
@@ -0,0 +1,91 @@
+/*
+ 	Copyright 2020 Google LLC
+
+ 	Licensed under the Apache License, Version 2.0 (the "License");
+ 	you may not use this file except in compliance with the License.
+ 	You may obtain a copy of the License at
+
+ 		https://www.apache.org/licenses/LICENSE-2.0
+
+ 	Unless required by applicable law or agreed to in writing, software
+ 	distributed under the License is distributed on an "AS IS" BASIS,
+ 	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ 	See the License for the specific language governing permissions and
+ 	limitations under the License.
+*/
+
+package com.google.play.core.godot.assetpacks.utils;
+
+import org.godotengine.godot.Dictionary;
+
+import com.google.android.play.core.assetpacks.AssetPackState;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Arrays;
+
+/**
+ * This class extends the AssetPackState abstract class, and provides constructor that allows
+ * PlayAssetDeliveryUtils class to instantiate AssetPackStateFromDictionary given Godot Dictionary.
+ */
+public class AssetPackStateFromDictionary extends AssetPackState {
+  private String name;
+  private int status;
+  private int errorCode;
+  private long bytesDownloaded;
+  private long totalBytesToDownload;
+  private int transferProgressPercentage;
+
+  public AssetPackStateFromDictionary(Dictionary dict)
+      throws NullPointerException, ClassCastException {
+    Set<String> dictionaryRequiredKeySet =
+        new HashSet<>(
+            Arrays.asList(
+                "name",
+                "status",
+                "errorCode",
+                "bytesDownloaded",
+                "totalBytesToDownload",
+                "transferProgressPercentage"));
+    if (dict.keySet().containsAll(dictionaryRequiredKeySet)) {
```

Comment:
You could get rid of this if-statement and rely on the casts to throw an exception if `dict.get("field")` returns null. You would have to check that `this.name` isn't null also (since `(String) null` won't throw an exception).

If you'd rather keep this check, I would make the field names static constants so that the string values aren't duplicated.

Reply:
I'm more in favor of the current if-statement check, since it is more modular and can be used by other util functions with little modification. 
The second solution will need more if-statements on Dictionary's with multiple String entries.

Reply:
Refactor this statement, added ```static``` and ```final``` keywords.

Reply:
I tried declaring public final Strings for the keys and referencing the declared Strings in the latest commit. In my opinion, this method can sacrifice some readability, especially for unfamiliar reader without IDEs, since the will have to figure out what is the actual value for key1. 
If we name the Strings with more context such as AssetPackStateFromDictionary. transferProgressPercentage, I think it is more straightforward to just use "transferProgressPercentage".
I'm kind of not sure about this decision, what is your opinion?

Reply:
You're right that naming them key1, key2 etc. reduces readability. I would definitely do what you suggested and give them more meaningful names like transferProgressPercentageKey.

You're right that it's more straightforward to just use the string, but if you do that, you sacrifice compiler checks on the key names. In other words, a typo in the field name would only be discovered at runtime. It also makes it clear to the reader, when they see these strings referenced in multiple functions, that those strings always need to match, rather than just coincidentally sharing a value. It also makes it easier to refactor if any of the strings change.

Reply:
Yup that make sense. Declaring those variables rather than repeating the same Strings multiple times in code does help with extending the code and reduce bugs.

Reply:
Thanks! Also I should have mentioned this in my last comment but constant variables should be named using `CONSTANT_CASE`, ex. `NAME_KEY`. See [here](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names)